Source/driver/Castro_advance.cpp

```diff
@@ -565,6 +597,202 @@ Castro::do_advance_mol (Real time,
+    }
+
+    // Now compute the advective term for the current node -- this
+    // will be used to advance us to the next node the new time
+
+    // Construct the primitive variables.
+    if (fourth_order) {
+      cons_to_prim_fourth(time);
+    } else {
+      cons_to_prim(time);
+    }
+
+    // Check for CFL violations.
+    check_for_cfl_violation(dt);
+
+    // If we detect one, return immediately.
+    if (cfl_violation)
+      return dt;
+
+    // construct the update for the current stage -- this fills
+    // A_new[m] with the righthand side for this stage.  Note, for m =
+    // 0, the starting state is S_old and never changes with SDC
+    // iteration, so we only do this once.
+    if (!(sdc_iteration > 0 && m == 0)) {
+      A_new[m]->setVal(0.0);
+      construct_mol_hydro_source(time, dt, *A_new[m]);
+    }
+
+    // also, if we are the first SDC iteration, we haven't yet stored
+    // any old advective terms, so we cannot yet do the quadrature
+    // over nodes.  Initialize those now.  Recall, A_new[0] and A_old[0]
+    // are aliased.
+    if (sdc_iteration == 0 && m == 0) {
+      for (int n=1; n < SDC_NODES; n++) {
+        MultiFab::Copy(*(A_old[n]), *(A_new[0]), 0, 0, NUM_STATE, 0);
+      }
+    }
+
+#ifdef REACTIONS
+    // if this is the first node of a new iteration, then we need
+    // to compute and store the old reactive source
+    if (m == 0 && sdc_iteration == 0) {
+      construct_old_react_source(Sborder, *(R_old[0]));
+
+      // copy to the other notes
+      for (int n = 1; n < SDC_NODES; n++) {
+        MultiFab::Copy(*(R_old[n]), *(R_old[0]), 0, 0, R_old[0]->nComp(), 0);
+      }
+    }
+#endif
+
+    // update to the next stage -- this involves computing the
+    // integral over the k-1 iteration data.  Note we don't do
+    // this if we are on the final node (since there is nothing to
+    // update to
+    if (m < SDC_NODES-1) {
+      do_sdc_update(m, m+1, dt); //(dt_sdc[m+1] - dt_sdc[m])*dt);
+    }
+
+  } // node iteration
+
+  // we are done with the integration over all nodes for this iteration
+
+  // note: as of right now, S_new is not updated with the state from
+  // the final time node.  This means we can still use S_new as
+  // "scratch" until we finally set it.
+
+  // store A_old for the next SDC iteration -- don't need to do n=0,
+  // since that is unchanged
+  for (int n=1; n < SDC_NODES; n++) {
+    MultiFab::Copy(*(A_old[n]), *(A_new[n]), 0, 0, NUM_STATE, 0);
+  }
+
+#ifdef REACTIONS
+  // we just did the update, so now recompute the "old" reactive
+  // source for the next SDC iteration.  We don't need to do this for
+  // m = 0, since that state never changes.
+
+  for (int m = 1; m < SDC_NODES; ++m) {
+    // use a temporary storage
+    // TODO: do we need a clean state here?
+    MultiFab::Copy(S_new, *(k_new[m]), 0, 0, S_new.nComp(), 0);
+    expand_state(Sborder, cur_time, -1, Sborder.nGrow());
+    construct_old_react_source(Sborder, *(R_old[m]));
```

Comment:
Since we already did this for the `sdc_iteration == 0` case in the main loop above, don't we only need to apply this step for `sdc_iteration > 0`?

Reply:
yes, I think you are right

Reply:
no... wait.  This is different.  Up above, we constructed the initial reactive source term that will be used for the integration over the temporal nodes during the update from SDC iteration 0 -> 1.  Now we are getting ready for the next SDC iteration, and we have new states on the temporal nodes m > 0, so we need to store the burn term for this state, which will be the initial burn for the next iteration.

Now... it may be that we don't need to do this extra step for the very _last_ SDC iteration.